### PR TITLE
Added winch :dizzy:

### PR DESCRIPTION
**Pull request by Bplotka**
Changes:
- Added proto for winch mapper config
- Improved kedge client tripper
- Added route mapper to lib/map + tests
- Improved protogen (I required that on my box)
- Added winch server
- Added winch docs
- Added winch+kedge integration test

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by Bplotka** [comment]
Still investigating why auth headers are not propagating properly

File: winch/integration_test.go [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+	pb "github.com/mwitkow/kedge/_protogen/winch/config"
+	"github.com/mwitkow/kedge/lib/map"
+	"github.com/mwitkow/kedge/winch"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+var (
+	backendResolutionDuration = 10 * time.Millisecond
```
Comment (kdima) [inline winch/integration_test.go]: not used

File: winch/integration_test.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+	// It does not make sense if kedge is not secure.
+	s.localSecureKedges.SetupKedges(s.T(), http2ServerTlsConfig, 3)
+
+	s.routes.Update(&pb.MapperConfig{
+		Routes: []*pb.Route{
+			{
+				Type: &pb.Route_Direct{
+					Direct: &pb.DirectRoute{
+						Key:      "resource1.ext.example.com",
+						KedgeUrl: "https://" + moveToLocalhost(s.localSecureKedges.listeners[0].Addr().String()),
```
Comment (kdima) [inline winch/integration_test.go]: why do we need to move to localhost? kedges are running on localhost already

File: winch/routes.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+type DynamicRoutes struct {
+	mu     sync.Mutex
+	routes []kedge_map.Route
+}
+
+func NewDynamicRoutes() *DynamicRoutes {
+	return &DynamicRoutes{}
+}
+
+func (r *DynamicRoutes) Update(config *pb.MapperConfig) error {
```
Comment (kdima) [inline winch/routes.go]: I don't think we need dynamic routes

File: winch/integration_test.go (hunk 2 above)
Comment (Bplotka) [inline winch/integration_test.go]: Look for the `moveToLocalhost` function - it just moved to `localhost` "domain" from `127.0.0.1` form (:

File: http/client/client.go [hunk 4]
```diff
@@ -22,10 +23,19 @@ func (t *tripper) RoundTrip(req *http.Request) (*http.Response, error) {
...
+
 	destURL, err := t.mapper.Map(host)
 	if err == kedge_map.ErrNotKedgeDestination {
 		return t.parent.RoundTrip(req)
 	}
+	if err != nil {
+		return nil, err
+	}
+
+	tags := http_ctxtags.ExtractInbound(req)
```
Comment (Bplotka) [inline http/client/client.go]: just to get tags? To be honest I just copied that from kedge

File: winch/integration_test.go (hunk 1 above)
Comment (Bplotka) [inline winch/integration_test.go]: +1

File: winch/routes.go (hunk 3 above)
Comment (Bplotka) [inline winch/routes.go]: Well, I would vote to keep it - might be useful for debug.

**Comment by Bplotka** [comment]
Got offline approval from @kdima

**Pull request merged**